examples/speculative/speculative.cpp

```diff
@@ -161,6 +219,11 @@ int main(int argc, char ** argv) {
         // sample n_draft tokens from the draft model picking the best token
         int n_past_cur = n_past_dft;
         for (int i = 0; i < n_draft; ++i) {
+            // remember the grammar state
+            if (grammar_dft != NULL) {
+                grammar_mem[i] = llama_grammar_copy(grammar_dft);
+            }
```

Comment:
I wonder if it's possible to scope `grammar_dft` to the drafting phase, copying it from `grammar_tgt` when you start drafting, and avoid the need for saving per-token state. The grammar state is entirely determined by the sequence of characters accepted. If I understand the flow here correctly, and drafting starts at the end of the token sequence accepted by the target model thus far, then it should be valid to just clone the target grammar for drafting.

Reply:
Great idea! Removed `grammar_mem` via [ba199d8](https://github.com/ggerganov/llama.cpp/pull/2991/commits/ba199d84096e447341017c06907749bd5888233e)